File: go/internal/e2e/llm_inference_consumer_cancel_e2e_test.go
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package e2e

import (
	"net/http"
	"sync/atomic"
	"testing"

	copilot "github.com/github/copilot-sdk/go"
	"github.com/github/copilot-sdk/go/internal/e2e/testharness"
)

type llmConsumerCancelHandler struct {
	inferenceAttempts atomic.Int32
}

func (h *llmConsumerCancelHandler) OnLlmRequest(req *copilot.LlmInferenceRequest) error {
	served, err := llmServiceNonInference(req)
	if err != nil {
		return err
	}
	if served {
		return nil
	}
	if !llmIsInferenceURL(req.URL) {
		return llmRespondBuffered(req, 200, http.Header{"content-type": {"application/json"}}, "{}")
	}

	// Consumer-initiated cancellation: the consumer's own upstream call was
	// aborted, so it tells the runtime to give up on this request. No response
	// head is ever produced; the runtime should see a transport failure rather
	// than hanging.
	llmDrainRequest(req)
	h.inferenceAttempts.Add(1)
	return req.ResponseBody.Error("upstream call aborted by consumer", "cancelled")
}

func TestLlmInferenceConsumerCancel(t *testing.T) {
	ctx := testharness.NewTestContext(t)
	handler := &llmConsumerCancelHandler{}
	client := newLlmClient(ctx, handler)
	t.Cleanup(func() { client.ForceStop() })

	if err := client.Start(t.Context()); err != nil {
		t.Fatalf("Failed to start client: %v", err)
	}

	session, err := client.CreateSession(t.Context(), &copilot.SessionConfig{
		OnPermissionRequest: copilot.PermissionHandler.ApproveAll,
	})
	if err != nil {
		t.Fatalf("Failed to create session: %v", err)
	}

	_, sendErr := session.SendAndWait(t.Context(), copilot.MessageOptions{Prompt: "Say OK."})
	_ = session.Disconnect()

	// The runtime reached the inference step and the consumer's cancellation
	// terminated it (rather than the runtime hanging).
	if handler.inferenceAttempts.Load() == 0 {
		t.Fatal("Expected the inference callback to be attempted")
	}
	if sendErr != nil && len(sendErr.Error()) == 0 {
		t.Fatal("Expected a non-empty error string when a failure surfaces")
	}
}

File: java/src/test/java/com/github/copilot/LlmInferenceConsumerCancelE2ETest.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------------------------------------------*/

package com.github.copilot;

import static com.github.copilot.LlmInferenceTestSupport.drainRequest;
import static com.github.copilot.LlmInferenceTestSupport.headers;
import static com.github.copilot.LlmInferenceTestSupport.isInferenceUrl;
import static com.github.copilot.LlmInferenceTestSupport.newLlmClient;
import static com.github.copilot.LlmInferenceTestSupport.respondBuffered;
import static com.github.copilot.LlmInferenceTestSupport.serviceNonInference;
import static com.github.copilot.LlmInferenceTestSupport.setupCapiAuth;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.github.copilot.rpc.MessageOptions;
import com.github.copilot.rpc.PermissionHandler;
import com.github.copilot.rpc.SessionConfig;

/**
 * Verifies that a consumer-initiated cancellation (the consumer's own upstream
 * call was aborted) terminates the request via a response error rather than
 * hanging the runtime.
 */
public class LlmInferenceConsumerCancelE2ETest {

    private static E2ETestContext ctx;

    @BeforeAll
    static void setup() throws Exception {
        ctx = E2ETestContext.create();
    }

    @AfterAll
    static void teardown() throws Exception {
        if (ctx != null) {
            ctx.close();
        }
    }

    private static final class ConsumerCancelHandler implements LlmInferenceProvider {

        private final AtomicInteger inferenceAttempts = new AtomicInteger();

        @Override
        public void onLlmRequest(LlmInferenceRequest req) throws Exception {
            if (serviceNonInference(req)) {
                return;
            }
            if (!isInferenceUrl(req.getUrl())) {
                respondBuffered(req, 200, headers("content-type", "application/json"), "{}");
                return;
            }

            // Consumer-initiated cancellation: the consumer's own upstream call was
            // aborted, so it tells the runtime to give up on this request. No response
            // head is ever produced; the runtime should see a transport failure rather
            // than hanging.
            drainRequest(req);
            inferenceAttempts.incrementAndGet();
            req.getResponseBody().error("upstream call aborted by consumer", "cancelled");
        }
    }

    @Test
    void surfacesConsumerInitiatedCancel() throws Exception {
        setupCapiAuth(ctx);
        ConsumerCancelHandler handler = new ConsumerCancelHandler();

        try (CopilotClient client = newLlmClient(ctx, handler)) {
            CopilotSession session = client
                    .createSession(new SessionConfig().setOnPermissionRequest(PermissionHandler.APPROVE_ALL)).get();

            try {
                session.sendAndWait(new MessageOptions().setPrompt("Say OK.")).get(60, TimeUnit.SECONDS);
            } catch (Exception ignored) {
                // Expected: the consumer cancelled the inference request.
            }
            session.close();
        }

        // The runtime reached the inference step and the consumer's cancellation
        // terminated it (rather than the runtime hanging).
        assertTrue(handler.inferenceAttempts.get() > 0, "Expected the inference callback to be attempted");
    }
}
